File: Backend/Services/UserService.cs
using Backend.Dtos;
using Backend.Models;
using Backend.Repositories.Interfaces;
using Backend.Services.Interfaces;

namespace Backend.Services;

public class UserService(IUserRepository userRepository, ITokenService tokenService) : IUserService
{
    private readonly IUserRepository _userRepository = userRepository;
    private readonly ITokenService _tokenService = tokenService;

    public async Task RegisterUser(AuthenticationDto register)
    {
        var usernameAlreadyExists = await _userRepository.CheckUserExistsByName(register.Username);

        if (usernameAlreadyExists)
        {
            throw new InvalidOperationException($"User with name {register.Username} already exists");
        }

        var user = new User
        {
            Id = Guid.NewGuid(),
            Username = register.Username,
            Password = HashUserPassword(register.Password),
            MealPrepInterval = 7
        };

        await _userRepository.RegisterUser(user);
    }

    public async Task<(RefreshToken, AccessTokenDto)> LoginUser(AuthenticationDto login)
    {
        var user = await _userRepository.GetUser(login.Username);

        if (user == null || !BCrypt.Net.BCrypt.Verify(login.Password, user.Password))
        {
            throw new UnauthorizedAccessException($"Invalid email or password");
        }

        var tokens = await _tokenService.GenerateNewTokens(user);

        return tokens;
    }

    public async Task LogoutUser(string refreshToken)
    {
        await _tokenService.RevokeTokenOnLogout(refreshToken);
    }

    public async Task<(RefreshToken, AccessTokenDto)> RefreshTokens(string oldRefreshToken)
    {
        var tokens = await _tokenService.RegenerateTokens(oldRefreshToken);

        return tokens;
    }

    private static string HashUserPassword(string password)
    {
        return BCrypt.Net.BCrypt.HashPassword(password);
    }
}

File: Frontend/src/api/services/userService.ts
import type { AccessToken } from "../../domain/models/accessToken";
import type { AuthRequest } from "../../domain/models/authRequest";
import BaseService from "./baseService";

class UserService extends BaseService {
  constructor() {
    super("api/user");
  }

  public async registerUser(authRequest: AuthRequest): Promise<void> {
    await this.post<AuthRequest, void>("auth/register", authRequest);
  }

  public async loginUser(authRequest: AuthRequest): Promise<AccessToken> {
    let accessToken = await this.post<AuthRequest, AccessToken>(
      "auth/login",
      authRequest
    );

    return accessToken;
  }

  public async refreshTokens(): Promise<AccessToken> {
    let accessToken = await this.post<void, AccessToken>("auth/refresh");

    return accessToken;
  }

  public async logoutUser(): Promise<void> {
    await this.post<void, void>("logout");
  }
}

const userService = new UserService();

export default userService;
